### DISCUSSION TITLE
TraderX deployment and orchestration

### DISCUSSION THREAD

**Discussion by maoo** [finos/traderX#190] [Ideas]
# Deploying TraderX Using Radius

![radius-diagram](https://github.com/user-attachments/assets/4c31e9c4-6ac3-4e22-9686-6beeea853ca2)

## 07/16/24: FINOS Tech Sprint 2024 Kick-off
[Watch the recording](https://www.finos.org/hubfs/Tech%20Sprint%202024/Tech%20Sprint%20Use%20Cases-%20TraderX%20deployment%20and%20orchestration%20%23190%20CORRECTED.mp4)

## Business Problem
TraderX is currently deployed using Docker Compose, which is an excellent tool for local development and testing. However, it lacks the scalability and robust orchestration capabilities required for large-scale production environments. This limitation hinders TraderX's ability to handle increased traffic and maintain high availability.

## Proposed Solution
Migrate TraderX from Docker Compose to Radius, a cloud-native application platform designed for scalability and portability. Radius, recently open-sourced and accepted as a CNCF sandbox project, offers a comprehensive solution for managing and orchestrating containerized applications.

## Roadmap
**"Radify" the components of TraderX**: 
- Configurable database and message bus resources 

**End-to-end automated CI/CD**: 
- Build a CI pipeline with GitHub Actions
- Integrate a GitOps platform (e.g. Flux, ArgoCD) for CD

**Multi-environment deployment**: 
- Reference architecture for deploying to local, Azure, AWS
- One-click deployment to experience TraderX on any cloud
- Developer dashboardd experience with application graph

## Related Technologies/Platforms
- Radius: The core platform for application deployment and management.
- Kubernetes: The underlying container orchestration engine.
- Docker: The containerization technology for packaging TraderX.
- Prometheus/Grafana: Potential tools for monitoring and observability.

## Key people helping to facilitate this hackathon stream: 
1. Jonathan Smith (@jonvsm) - Group Program Manager - Microsoft Azure Incubations
2. Will Tsai (@willtsai) - Principal Product Manager - Microsoft Azure Incubations

## More info on Radius: 
- Radius was open sourced under the Apache 2.0 license in October 2023 - [Introducing Radius, a new open-source application platform for cloud-native apps | Radius Blog (radapp.io)](https://blog.radapp.io/posts/2023/10/18/introducing-radius-a-new-open-source-application-platform-for-cloud-native-apps/)
- Radius was approved as a CNCF sandbox project in April 2024 - [Radius Accepted as Cloud Native Compute Foundation (CNCF) Sandbox Project | Radius Blog (radapp.io)](https://blog.radapp.io/posts/2024/04/16/radius-accepted-as-cloud-native-compute-foundation-cncf-sandbox-project/)
- Millenium BCP, the largest bank in Portugal, published a case study describing their use of Radius - [Case Study: How Millennium bcp leverages Radius | Radius Blog (radapp.io)](https://blog.radapp.io/posts/2023/12/06/case-study-how-millennium-bcp-leverages-radius/)
- Radius Docs - [Radius (radapp.io)](https://radapp.io/)
- Radius on GitHub - [Radius (github.com)](https://github.com/radius-project)

**Comment by rocketstack-matt** [comment]
I think there could be an opportunity here to use [CALM](https://github.com/finos/architecture-as-code/tree/main/calm) to model TraderX (also see [this discussion](https://github.com/finos/traderX/discussions/192)) and map it to the GitOps tooling @caldeirav has built and uses with OS-Climate to drive the orchestration. 

What do you think @caldeirav?

CC: @dc-ms / @yt-ms

**Reply by caldeirav** [reply]
We could definitely support declarative management of application definition, application configuration and environment and we typically leverage for this ArgoCD (https://argoproj.github.io/cd/) - it is also a CNCF project and is mature for your continuous delivery requirements. Integration with Radius will have to be evaluated. 

In terms of application service discovery, routing and load balancing, Kubernetes is actually evolving towards a new generation of Ingress, Load Balancing and Service Mesh APIs which is Kubernetes Gateway API (https://gateway-api.sigs.k8s.io). The Red Hat team could potentially support you on the implementation. We have an upstream project, Kuadrant (https://kuadrant.io) that is geared towards making the management of such behaviors directly accessible and easy for developers, and will integrate natively with a GitOps approach.

**Comment by lucaborella89** [comment]
Hi @jonvsm I have updated the description based on what we discussed. Feel free to make suggestion in this comment and will edit it again. Thanks

**Reply by refset** [reply]
> [jonvsm](https://github.com/jonvsm)
[Jonathan Smith](https://github.com/jonvsm)

You have the wrong name under 'Key people' (showing 'Jeremy Taylor' still :slightly_smiling_face:)
